server/src/routes/slackbot.js

```diff
@@ -0,0 +1,123 @@
+import express from 'express'
+import request from 'request'
+import { uuid, reqUser } from './helpers'
+import { catchErrors } from '../handlers'
+import {
+  addSlackUser,
+  linkSlackToUser,
+  setSlackUser,
+  getSlackUser,
+  confirmContainerOwnership,
+  isLinkedToSlack,
+} from '../database'
+import { config } from '../globals'
+
+const router = express.Router()
+
+// CHANGE THESE TO MATCH USED SLACK APP IN ../../local.json
+
+// @api GET /api/slack/oauth
+// Authentication with Slack sign in.
+// This path should only be called by Slack oauth after pressing 'Sign in with Slack'-button.
+//
+// Returns error if authentication failed or redirects back to dashboard otherwise
+router.get('/oauth', catchErrors(async (req, res) => {
+  const { clientId, clientSecret } = config.slack || {}
+  const { state } = req.query
+  const options = {
+    uri: `https://slack.com/api/oauth.access?code=${req.query.code}&client_id=${clientId}+'&client_secret=${clientSecret}`,
+    method: 'GET',
+  }
+  await request(options, async (error, response, body) => {
+    const parsedBody = JSON.parse(body)
+    if (!parsedBody.ok) {
+      res.send(`Error encountered: \n${JSON.stringify(parsedBody)}`)
+        .status(200)
+        .end()
+    } else {
+      await setSlackUser(state, parsedBody.user.name, parsedBody.user.id)
+      res.redirect('/')
+    }
+  })
+}))
+
+// @api POST /api/slack/oauth/connect
+// For letting our slack authentication know who user clicked 'Sign in with Slack'-button
+//
+// Returns redirect to Slack's oauth.
+router.post('/oauth/connect', catchErrors(async (req, res) => {
+  const { clientId } = config.slack || {}
+  const { users } = await reqUser(req)
+  const id = uuid(8)
+  await addSlackUser(id)
+  for (const userId of Object.keys(users)) {
+    await linkSlackToUser(id, userId)
+  }
+  const slackURL = `https://slack.com/oauth/authorize?scope=identity.basic&client_id=${clientId}`
+  res.json({ url: `${slackURL}&state=${id}` })
+}))
+
+// @api GET /api/slack/auth
+// For checking if user has connected to Slack
+//
+// Returns json containing boolean indicating whether connected or not.
+// Contains slack username and user id if connected as well.
+router.get('/auth', catchErrors(async (req, res) => {
+  const { users } = await reqUser(req)
+  for (const userId of Object.keys(users)) {
+    const slackUser = await getSlackUser(userId)
+    if (slackUser.length > 0 && slackUser[0].slack_user_id) {
```

Comment:
what happens with multiple users

Reply:
it checks if atleast one of users is connected to slack